packages/flutter_tools/lib/src/plugins.dart

```diff
@@ -351,21 +395,165 @@ Future<List<Plugin>> findPlugins(FlutterProject project, { bool throwOnError = t
     logger: globals.logger,
     throwOnError: throwOnError,
   );
   for (final Package package in packageConfig.packages) {
     final Uri packageRoot = package.packageUriRoot.resolve('..');
-    final Plugin plugin = _pluginFromPackage(package.name, packageRoot);
+    final Plugin plugin = _pluginFromPackage(
+      package.name,
+      packageRoot,
+      <String>[...appDependencies.keys.cast<String>()],
+    );
     if (plugin != null) {
       plugins.add(plugin);
     }
   }
   return plugins;
 }
 
+/// Metadata associated with the resolution of a platform interface of a plugin.
+class PluginInterfaceResolution {
+  PluginInterfaceResolution({
+    @required this.plugin,
+    this.platform,
+    this.dartClass,
+  }) : assert(plugin != null);
+
+  /// The plugin.
+  final Plugin plugin;
+  // The name of the platform that this plugin implements.
+  final String platform;
+  // The name of the Dart class.
+  final String dartClass;
+
+  Map<String, String> toMap() {
+    return <String, String> {
+      'pluginName': plugin.name,
+      'dartClass': dartClass,
+      'platform': platform,
+    };
+  }
+}
+
+/// Resolves the platform implementation for Dart-only plugins.
+///
+///   * If there are multiple direct pub dependencies on packages that implement the
+///     frontend plugin for the current platform, fail.
+///   * If there is a single direct dependency on a package that implements the
+///     frontend plugin for the target platform, this package is the selected implementation.
+///   * If there is no direct dependency on a package that implements the frontend
+///     plugin for the target platform, and the frontend plugin has a default implementation
+///     for the target platform the default implementation is selected.
+///   * Else fail.
+///
+///  For more details, https://flutter.dev/go/federated-plugins.
+List<PluginInterfaceResolution> resolvePlatformImplementation(
+  List<Plugin> plugins, {
+  bool throwOnPluginPubspecError = true,
+}) {
+  final List<String> platforms = <String>[
+    LinuxPlugin.kConfigKey,
+    MacOSPlugin.kConfigKey,
+    WindowsPlugin.kConfigKey,
+  ];
+  final Map<String, PluginInterfaceResolution> directDependencyResolutions
+      = <String, PluginInterfaceResolution>{};
+
+  final Map<String, String> defaultImplementations = <String, String>{};
+
+  for (final Plugin plugin in plugins) {
+    for (final String platform in platforms) {
+      // The plugin doesn't implement this platform.
+      if (plugin.platforms[platform] == null) {
+        continue;
+      }
+      final Map<String, dynamic> platformEntries = plugin.platforms[platform].toMap();
+      // The plugin doesn't implement an interface, verify that it has a default implementation.
+      if (plugin.implement == null || plugin.implement.isEmpty) {
+        final String defaultImplementation = platformEntries[kDefaultPackage] as String;
+        if (defaultImplementation == null) {
+          if (throwOnPluginPubspecError) {
```

Comment:
In the event that the user as N plugins that all have this same problem, immediately throwing the error will force them to resolve them one at a time. The pattern used in the asset manifest and pubspec validation is to printError and continue to the end, throwing if you've encountered at least one error.

Reply:
Done